addons/utm/tests/test_utm.py

```diff
@@ -0,0 +1,63 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.addons.utm.tests.common import TestUTMCommon
+
+
+class TestUtm(TestUTMCommon):
+    def test_find_or_create_record(self):
+        source_1, source_2 = self.env['utm.source'].create([{
+            'name': 'Source 1',
+        }, {
+            'name': 'Source 2',
+        }])
+
+        # Find the record based on the given name
+        source = self.env['utm.mixin']._find_or_create_record('utm.source', 'Source 1')
+        self.assertEqual(source, source_1)
+
+        # Create a new record
+        source_4 = self.env['utm.mixin']._find_or_create_record('utm.source', 'Source 3')
+        self.assertNotIn(source_4, source_1 | source_2)
+        self.assertEqual(source_4.name, 'Source 3')
+
+    def test_name_generation(self):
+        """Test that the name is always unique.
+
+        A counter must be added at the end of the name if it's not the case.
+        """
+        for utm_model in ('utm.source', 'utm.medium', 'utm.campaign'):
+            utm_0 = self.env[utm_model].create({'name': 'UTM dup'})
+
+            utm_1, utm_2, utm_3, utm_4, utm_5 = self.env[utm_model].create([{
+                'name': 'UTM 1',
+            }, {
+                'name': 'UTM 2',
+            }, {
+                'name': 'UTM dup',
+            }, {
+                # UTM record 4 has the same name of the previous UTM record
+                'name': 'UTM dup',
+            }, {
+                # UTM record 5 has the same name of the previous UTM record
+                # but with a wrong counter part, it will be removed and updated
+                'name': 'UTM dup [0]',
+            }])
+
+            self.assertEqual(utm_0.name, 'UTM dup', msg='Counter "[1]" should not be added')
```

Comment:
'Counter "[1]" should not be added'

You can't be 100% sure that this is going to be the issue if someone breaks your method one day.

I would go for "The first 'UTM dup' should be left unchanged since it's unique" or something like that.